staging/src/k8s.io/apiserver/pkg/cel/common/valuesreflect.go

```diff
@@ -0,0 +1,678 @@
+				if schema, ok := schema.Properties()[key]; ok {
+					return schema, true
+				}
+				return nil, false
+			},
+		}
+	// Match type aliases to primitives by kind
+	case reflect.Bool:
+		return types.Bool(v.Bool())
+	case reflect.String:
+		return types.String(v.String())
+	case reflect.Int, reflect.Int32, reflect.Int64:
+		return types.Int(v.Int())
+	case reflect.Float32, reflect.Float64:
+		return types.Double(v.Float())
+	default:
+		return types.NewErr("unsupported Go type for CEL: %v", v.Type())
+	}
+}
+
+// typedStruct wraps a struct as a CEL ref.Val and provides lazy access to fields via reflection.
+type typedStruct struct {
+	value reflect.Value // Kind is required to be: reflect.Struct
+
+	// propSchema finds the schema to use for a particular map key.
+	propSchema func(key string) (Schema, bool)
+}
+
+func (s *typedStruct) ConvertToNative(typeDesc reflect.Type) (interface{}, error) {
+	if s.value.Type().AssignableTo(typeDesc) {
+		return s.value.Interface(), nil
+	}
+	return nil, fmt.Errorf("type conversion error from struct type %v to %v", s.value.Type(), typeDesc)
+}
+
+func (s *typedStruct) ConvertToType(typeValue ref.Type) ref.Val {
+	switch typeValue {
+	case s.Type():
+		return s
+	case types.MapType:
+		return s
+	case types.TypeType:
+		return s.objType()
+	}
+	return types.NewErr("type conversion error from struct %s to %s", s.Type().TypeName(), typeValue.TypeName())
+}
+
+func (s *typedStruct) Equal(other ref.Val) ref.Val {
+	otherStruct, ok := other.(*typedStruct)
+	if ok {
+		return types.Bool(apiequality.Semantic.DeepEqual(s.value.Interface(), otherStruct.value.Interface()))
+	}
+	return types.MaybeNoSuchOverloadErr(other)
+}
+
+func (s *typedStruct) Type() ref.Type {
+	return s.objType()
+}
+
+func (s *typedStruct) objType() *types.Type {
+	typeName := s.value.Type().Name()
+	if pkgPath := s.value.Type().PkgPath(); pkgPath != "" {
+		typeName = pkgPath + "." + typeName
+	}
+	return types.NewObjectType(typeName)
+}
+
+func (s *typedStruct) Value() interface{} {
+	return s.value.Interface()
+}
+
+func (s *typedStruct) IsSet(field ref.Val) ref.Val {
+	v, found := s.lookupField(field)
+	if v != nil && types.IsUnknownOrError(v) {
+		return v
+	}
+	return types.Bool(found)
+}
+
+func (s *typedStruct) Get(key ref.Val) ref.Val {
+	v, found := s.lookupField(key)
+	if !found {
+		return types.NewErr("no such key: %v", key)
+	}
+	return v
+}
+
+func (s *typedStruct) lookupField(key ref.Val) (ref.Val, bool) {
+	keyStr, ok := key.(types.String)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(key), true
+	}
+	fieldName := keyStr.Value().(string)
+
+	cacheEntry := value.TypeReflectEntryOf(s.value.Type())
+	fieldCache, ok := cacheEntry.Fields()[fieldName]
+	if !ok {
+		return nil, false
+	}
+
+	if e := fieldCache.GetFrom(s.value); !fieldCache.CanOmit(e) {
+		if propSchema, ok := s.propSchema(fieldName); ok {
+			v := TypedToVal(e.Interface(), propSchema)
+			if v == types.NullValue {
```

Comment:
If a field is nil, should it count as finding the value?

If I use the CEL playground:

```
has(spec.explicitlyNullField)
// has(spec.absentField)
```

```
spec:
  explicitlyNullField: null
  # The'absentField' missing spec
```

I get:

has(spec.explicitlyNullField) -> true
has(spec.absentField) -> false

Would this implementation also return this?  

https://playcel.undistro.io/?content=H4sIAAAAAAAAA6tWSk7NUbJSykgs1iguSE3WS60oyMlMzizJqfQrzclxy0zNSdGMydPXV4CrSEwqTs0rgcgo6SilJJYkeuYVlJYATQHJW8XkKShgMcVKIQ%2FIBEkqK4RkpKojGaOukJtZXJyZl64AMgBoZm5%2BSirQOJDTagGIfws3ogAAAA%3D%3D

Reply:
https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros is our source of truth.  I don't think the playground is doing the right thing here.  CRD validation rules operate on the semantic of "A field is available if it's value is present (non-nil)".  This interacts well with users that guard dereferences with `has()` checks, so I'd like to continue with that approach.

Reply:
in places where we're switching from `tounstructured()` to this (https://github.com/kubernetes/kubernetes/issues/131520), let's spot test the behavior of `has()` doesn't change

I would *guess* that places where we were omitting null stanzas / attributes ([example](https://github.com/kubernetes/kubernetes/blob/release-1.33/staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go#L282)) will work the same, but we should verify that

Reply:
Yes, the test suite in use here tests both the unstructured and this new "reflective" approach and ensure they behave the same, for the test cases we have.  I'll do a sweep specifically around has() and different stanza representations (pointer/non-pointer,zero-valued,embedded,inlined...)

Reply:
Did more has() testing. https://github.com/kubernetes/kubernetes/pull/130989 will help. I've verified that once that PR merges and `omitzero` both unstructured and typed CEL wrappers will correctly respect omitzero.

I'll merge additional tests once https://github.com/kubernetes/kubernetes/pull/130989 to demonstrate this.